# Analyzing High-Speed Railway Delays Project

- [View Solution Notebook](./solution.html)
- [View Project Page](https://www.codecademy.com/projects/practice/analyzing-high-speed-railways-with-python)

## Task Group 1 - Load and Clean

In [1]:
# Setup
import pandas as pd

df0 = pd.read_csv('datasets/railway_delays_2020_01_1.csv')
df0.head()

date           ride_id train_number  station_order  \
0  2020-01-01  G1226_20200101_1        G1226            1.0   
1  2020-01-01  G1226_20200101_2        G1226            2.0   
2  2020-01-01  G1226_20200101_3        G1226            3.0   
3  2020-01-01  G1226_20200101_4        G1226            4.0   
4  2020-01-01  G1226_20200101_5        G1226            5.0   

                  station_name                  to_station  mileage  \
0  Shenyangbei Railway Station    Shenyang Railway Station      5.0   
1     Shenyang Railway Station    Liaoyang Railway Station     65.0   
2     Liaoyang Railway Station    Anshanxi Railway Station     30.0   
3     Anshanxi Railway Station  Haichengxi Railway Station     39.0   
4   Haichengxi Railway Station      Panjin Railway Station     66.0   

   arrival_delay  departure_delay  major_holiday  
0            0.0             -1.0           True  
1           -2.0              2.0           True  
2            4.0              4.0           True  
3            5.0              4.0           True  
4            5.0              4.0           True

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What is the structure of this dataset? Toggle to check!</i></summary>

Each row contains departure and arrival delays for trains in China on January 1st, 2020. 

Here's a quick summary of the columns:

- **date** - the date of the train ride
- **ride_id** - the unique identifier for each train ride: `<train_number>_<yyyymmdd>_<station_order>`
- **train_number** - a unique identifier of the train
- **station_order** - the order in which each station was visited in a train route by a single train number on a specific date
- **station_name** - the name of the departing station
- **to_station** - the name of the next station destination
- **mileage** - the distance in kilometers to the next station
- **arrival_delay** - the delay in minutes arriving to the next station
- **departure_delay** - the delay in minutes departing the current station
- **major_holiday** - `True` if the `date` is a major holiday

</details>

### Task 1

Each day in January has its own data file. The names of these files depend on the date. For example, the file we read in above is `datasets/railway_delays_2020_01_1.csv`, corresponding to the day `2020-01-1`, or January 1st.

Use a list comprehension to create a list of the filenames from January 1st, 2020 to January 27th, 2020. Name this list `filenames` and print out the list. 

In [3]:
filenames = ['datasets/railway_delays_2020_01_'+str(i)+'.csv' for i in range(1, 28)]
print(filenames)

['datasets/railway_delays_2020_01_1.csv', 'datasets/railway_delays_2020_01_2.csv', 'datasets/railway_delays_2020_01_3.csv', 'datasets/railway_delays_2020_01_4.csv', 'datasets/railway_delays_2020_01_5.csv', 'datasets/railway_delays_2020_01_6.csv', 'datasets/railway_delays_2020_01_7.csv', 'datasets/railway_delays_2020_01_8.csv', 'datasets/railway_delays_2020_01_9.csv', 'datasets/railway_delays_2020_01_10.csv', 'datasets/railway_delays_2020_01_11.csv', 'datasets/railway_delays_2020_01_12.csv', 'datasets/railway_delays_2020_01_13.csv', 'datasets/railway_delays_2020_01_14.csv', 'datasets/railway_delays_2020_01_15.csv', 'datasets/railway_delays_2020_01_16.csv', 'datasets/railway_delays_2020_01_17.csv', 'datasets/railway_delays_2020_01_18.csv', 'datasets/railway_delays_2020_01_19.csv', 'datasets/railway_delays_2020_01_20.csv', 'datasets/railway_delays_2020_01_21.csv', 'datasets/railway_delays_2020_01_22.csv', 'datasets/railway_delays_2020_01_23.csv', 'datasets/railway_delays_2020_01_24.csv', 

### Task 2

Now that we have a list of filenames, we can import them as DataFrames. Use a list comprehension to create a list of DataFrames named `railway_delays_dfs`. Print summary information about the first DataFrame and take a look at the column names and their datatypes.

In [5]:
railway_delays_dfs = [pd.read_csv(file) for file in filenames]
railway_delays_dfs[0].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1111 entries, 0 to 1110
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   date             1111 non-null   object 
 1   ride_id          1111 non-null   object 
 2   train_number     1111 non-null   object 
 3   station_order    1111 non-null   float64
 4   station_name     1111 non-null   object 
 5   to_station       1111 non-null   object 
 6   mileage          1111 non-null   float64
 7   arrival_delay    1111 non-null   float64
 8   departure_delay  1111 non-null   float64
 9   major_holiday    1111 non-null   bool   
dtypes: bool(1), float64(4), object(5)
memory usage: 79.3+ KB


### Task 3

Concatenate the DataFrames in `railway_delays_dfs` into a single DataFrame named `railway_delays_full`. Print summary information about `railway_delays_full`.

What do you notice about the columns in our fully concatenated DataFrame?

In [7]:
railway_delays_full = pd.concat(railway_delays_dfs)
railway_delays_full.info

<bound method DataFrame.info of             date            ride_id train_number  station_order  \
0     2020-01-01   G1226_20200101_1        G1226            1.0   
1     2020-01-01   G1226_20200101_2        G1226            2.0   
2     2020-01-01   G1226_20200101_3        G1226            3.0   
3     2020-01-01   G1226_20200101_4        G1226            4.0   
4     2020-01-01   G1226_20200101_5        G1226            5.0   
...          ...                ...          ...            ...   
1077  2020-01-27  G7592_20200127_20        G7592           20.0   
1078  2020-01-27  G7592_20200127_21        G7592           21.0   
1079  2020-01-27  G7592_20200127_22        G7592           22.0   
1080  2020-01-27  G7592_20200127_23        G7592           23.0   
1081  2020-01-27  G7592_20200127_24        G7592           24.0   

                      station_name                    to_station  mileage  \
0      Shenyangbei Railway Station      Shenyang Railway Station      5.0   
1        

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What happened when we concatenated all of the DataFrames? Toggle to check!</i></summary>

Vertically concatenating all of the DataFrames in `railway_delays_dfs` resulted in duplicate columns. The `.info()` output tells us that each column name appears at least twice. One set of names is lowercased and the other is uppercased. We also see the final column `station_name` is another duplicate that probably resulted from the presence of extra whitespaces in its name.

These discrepancies are most likely due to individual DataFrames containing a different set of column names. Since Python and pandas are case-sensitive, we'll need to standardize the column names in every DataFrame before concatenation. 
</details>

### Task 4

Standardize the column names in each of the railway delay DataFrames by 

- *lowercasing* each column name
- *removing* any extra whitespaces

Use a `for` loop to automate this process.

After standardizing the names, re-concatenate the cleaned DataFrames into a single DataFrame named `railway_delays_full` and preview the first five rows.

In [9]:
for df in railway_delays_dfs:
    df.columns = df.columns.str.lower()
    df.columns = df.columns.str.strip()
    
railway_delays_full = pd.concat(railway_delays_dfs)
railway_delays_full.head()

date           ride_id train_number  station_order  \
0  2020-01-01  G1226_20200101_1        G1226            1.0   
1  2020-01-01  G1226_20200101_2        G1226            2.0   
2  2020-01-01  G1226_20200101_3        G1226            3.0   
3  2020-01-01  G1226_20200101_4        G1226            4.0   
4  2020-01-01  G1226_20200101_5        G1226            5.0   

                  station_name                  to_station  mileage  \
0  Shenyangbei Railway Station    Shenyang Railway Station      5.0   
1     Shenyang Railway Station    Liaoyang Railway Station     65.0   
2     Liaoyang Railway Station    Anshanxi Railway Station     30.0   
3     Anshanxi Railway Station  Haichengxi Railway Station     39.0   
4   Haichengxi Railway Station      Panjin Railway Station     66.0   

   arrival_delay  departure_delay  major_holiday  
0            0.0             -1.0           True  
1           -2.0              2.0           True  
2            4.0              4.0           True  
3            5.0              4.0           True  
4            5.0              4.0           True

### Task 5
Let's continue cleaning the DataFrame. Use a for loop and `if/elif` statements to

- *uppercase* all the object datatype columns
- *strip* any hidden extra whitespace from object columns
- *convert* all float datatype columns to integers

Preview the cleaned DataFrame to confirm the transformations.

In [14]:
for columns in railway_delays_full.columns:
    if railway_delays_full[columns].dtype == 'object':
        railway_delays_full[columns] = railway_delays_full[columns].str.upper()
    elif railway_delays_full[columns].dtype == 'float64':
        railway_delays_full[columns] = railway_delays_full[columns].astype('int64')
        
railway_delays_full.head()

date           ride_id train_number  station_order  \
0  2020-01-01  G1226_20200101_1        G1226              1   
1  2020-01-01  G1226_20200101_2        G1226              2   
2  2020-01-01  G1226_20200101_3        G1226              3   
3  2020-01-01  G1226_20200101_4        G1226              4   
4  2020-01-01  G1226_20200101_5        G1226              5   

                  station_name                  to_station  mileage  \
0  SHENYANGBEI RAILWAY STATION    SHENYANG RAILWAY STATION        5   
1     SHENYANG RAILWAY STATION    LIAOYANG RAILWAY STATION       65   
2     LIAOYANG RAILWAY STATION    ANSHANXI RAILWAY STATION       30   
3     ANSHANXI RAILWAY STATION  HAICHENGXI RAILWAY STATION       39   
4   HAICHENGXI RAILWAY STATION      PANJIN RAILWAY STATION       66   

   arrival_delay  departure_delay  major_holiday  
0              0               -1           True  
1             -2                2           True  
2              4                4           True  
3              5                4           True  
4              5                4           True

## Task Group 2 - Impact of Holiday Travel

### Task 6

Let's see if major holidays had an impact on railway delay times.

Let's first see how many train rides actually occurred on major holidays compared to those that didn't. Use the `.value_counts()` method on the `major_holiday` column in `railway_delays_full`.

There are four major holidays in our dataset. How does the number of trains on holidays compare to the number of trains on non-holidays?

In [16]:
railway_delays_full['major_holiday'].value_counts()

False    25443
True      4274
Name: major_holiday, dtype: int64

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What did we discover in Task 6? Toggle to check!</i></summary>

There are `4274` individual station->station trips on the major holidays, averaging to `1068` per day.
    
On the `23` non-holiday days in our dataset, there are `25443` station->station trips, averaging to `1106` per day.
    
Based on this rough back-of-the-envelope calculation, there do seem to be slightly fewer individual trips on the holidays. But we'd have to look at data for more months or talk to a subject-matter expert to know for sure.
</details>

### Task 7

Groupby `major_holiday` and calculate the average of `arrival_delay` and `departure_delay`.

In [18]:
railway_delays_full.groupby('major_holiday').agg({'arrival_delay':'mean', 'departure_delay':'mean'})

arrival_delay  departure_delay
major_holiday                                
False              23.367881        22.362104
True               23.815863        22.924193

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What did we discover about the impact of holiday dates? Toggle to check!</i></summary>

There doesn't seem to be a major difference in average delays, though both are slightly higher on holidays than non-holidays (by less than a minute). We'd have to dig into this further to know if this is a fluke of the dates we happen to be looking at or an actual statistical impact. Either way, it probably isn't enough to make a noticeable difference to riders!

</details>

## Task Group 3 - Impact of Distance Between Stations

### Task 8
The distance between stations could impact delays. For example, having further to travel might correspond to increased delays.

In the next tasks, we'll investigate the distances between stations to see if there is an impact on train delays. 

Compute descriptive summary statistics on the distances from station to station by calling the `.describe()` method on the `mileage` column in `railway_delays_full`.

What is the mean mileage (kilometers)?

In [20]:
railway_delays_full['mileage'].describe()

count    29717.000000
mean        88.000236
std         79.537992
min          1.000000
25%         43.000000
50%         70.000000
75%        120.000000
max       1410.000000
Name: mileage, dtype: float64

### Task 9

Let's see if longer-than-average distances correspond to longer delays.

Create a column `long_distance` that is `True` in any row with `mileage` greater than `88` (the average).

In [22]:
railway_delays_full['long_distance'] = railway_delays_full['mileage'] > 88

railway_delays_full[['mileage', 'long_distance']].head()

mileage  long_distance
0        5          False
1       65          False
2       30          False
3       39          False
4       66          False

### Task 10

Groupby `long_distance` and calculate the average `arrival_delay` and `departure_delay`.

In [25]:
railway_delays_full.groupby('long_distance').agg({'arrival_delay':'mean', 'departure_delay':'mean'})

arrival_delay  departure_delay
long_distance                                
False              16.578222        15.120508
True               34.611244        34.385746

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What did we discover about the impact of distances between stations? Toggle to check!</i></summary>

Trips with longer-than-average distance have almost double the mean delays of shorter-than-average trips. While we can't be 100% certain that the distances from station to station are a contributing factor to delay times, this is promising evidence to investigate this phenomenon further. 

</details>

## Task Group 4 - Impact of Weather

### Task 11

Lastly, let's take a look at delays during ideal weather conditions and compare them to the delays during severe weather conditions.

Import the CSV file `datasets/weather.csv` into a DataFrame named `weather` and preview the first five rows.

In [27]:
weather = pd.read_csv('datasets/weather.csv')
weather.head()

ride_id                      wind weather  temperature
0  G1226_20200101_1    light winds from the N   sunny         -7.0
1  G1226_20200101_2    light winds from the N   sunny         -7.0
2  G1226_20200101_3    light winds from the N   sunny         -3.0
3  G1226_20200101_4  gentle breeze from the N   sunny         -3.0
4  G1226_20200101_5  gentle breeze from the N   sunny         -3.0

### Task 12

Let's clean `weather` by 
- *uppercasing* all of the text/object datatype columns
- *stripping* whitespace from the object columns
- *converting* all the float datatype columns to integers (we've checked that they should all be integers!) 

Use `for` loops and `if/elif` statements to automate this process.

Preview the modified DataFrame. 

In [28]:
for column in weather.columns:
    if weather[column].dtype == 'object':
        weather[column] = weather[column].str.upper()
        weather[column] = weather[column].str.strip()
    
    elif weather[column].dtype == 'float64':
        weather[column] = weather[column].astype('int64')

weather.head()

ride_id                      wind weather  temperature
0  G1226_20200101_1    LIGHT WINDS FROM THE N   SUNNY           -7
1  G1226_20200101_2    LIGHT WINDS FROM THE N   SUNNY           -7
2  G1226_20200101_3    LIGHT WINDS FROM THE N   SUNNY           -3
3  G1226_20200101_4  GENTLE BREEZE FROM THE N   SUNNY           -3
4  G1226_20200101_5  GENTLE BREEZE FROM THE N   SUNNY           -3

### Task 13

Merge `weather` with `railway_delays_full` using the `ride_id` column. Use an inner merge and name the merged DataFrame `railway_delays_join_weather`. Preview the first five rows of the DataFrame.

In [31]:
railway_delays_join_weather = pd.merge(left = weather,
                                      right = railway_delays_full,
                                      left_on = 'ride_id',
                                      right_on = 'ride_id',
                                      how = 'inner')
railway_delays_join_weather.head()

ride_id                      wind weather  temperature  \
0  G1226_20200101_1    LIGHT WINDS FROM THE N   SUNNY           -7   
1  G1226_20200101_2    LIGHT WINDS FROM THE N   SUNNY           -7   
2  G1226_20200101_3    LIGHT WINDS FROM THE N   SUNNY           -3   
3  G1226_20200101_4  GENTLE BREEZE FROM THE N   SUNNY           -3   
4  G1226_20200101_5  GENTLE BREEZE FROM THE N   SUNNY           -3   

         date train_number  station_order                 station_name  \
0  2020-01-01        G1226              1  SHENYANGBEI RAILWAY STATION   
1  2020-01-01        G1226              2     SHENYANG RAILWAY STATION   
2  2020-01-01        G1226              3     LIAOYANG RAILWAY STATION   
3  2020-01-01        G1226              4     ANSHANXI RAILWAY STATION   
4  2020-01-01        G1226              5   HAICHENGXI RAILWAY STATION   

                   to_station  mileage  arrival_delay  departure_delay  \
0    SHENYANG RAILWAY STATION        5              0               -1   
1    LIAOYANG RAILWAY STATION       65             -2                2   
2    ANSHANXI RAILWAY STATION       30              4                4   
3  HAICHENGXI RAILWAY STATION       39              5                4   
4      PANJIN RAILWAY STATION       66              5                4   

   major_holiday  long_distance  
0           True          False  
1           True          False  
2           True          False  
3           True          False  
4           True          False

### Task 14

Groupby the `weather` column, and calculate the average of `arrival_delay` and `departure_delay`. Sort the result from highest `arrival_delay` to lowest.

In [35]:
railway_delays_join_weather.groupby('weather').agg({'arrival_delay':'mean', 'departure_delay':'mean'}).sort_values('arrival_delay', ascending = False)

arrival_delay  departure_delay
weather                                               
BLIZZARD                   314.666667       314.666667
HEAVY SNOW                  79.289157        78.590361
MODERATE SNOW               67.816794        66.977099
LIGHT SNOW                  50.243197        49.719388
SLEET                       39.614767        39.579454
MODERATE RAIN               39.065776        37.885920
CLOUDY                      28.314811        27.243096
SUNNY                       23.074158        21.916419
OVERCAST                    19.193166        18.550558
LIGHT RAIN                  15.308342        14.110996
SHOWERS                      7.551724         7.195402
SNOW SHOWERS                 3.527027         3.229730
LIGHT TO MODERATE RAIN       2.960000         1.184615
DOWNPOUR                     2.571429         1.771429
MODERATE TO HEAVY RAIN       0.833333         0.333333
FOG                          0.689922         0.031008
MODERATE HAZE                0.650943         0.490566
HAZE                         0.501946         0.163424
HEAVY HAZE                  -0.125000        -0.166667

<details>
    <summary style="display:list-item; font-size:16px; color:blue;"><i>What did we discover about the impact of weather? Toggle to check!</i></summary>

There is a noticeable difference in delays based on weather conditions.

- Severe weather conditions like blizzards and heavy snow saw a significant increase in delay times
- Ideal weather conditions like sunny and rainy conditions saw shorter delay times
- Fog and hazy weather conditions had the lowest delay times

The last bullet point is perhaps the most interesting. Can you think of any reasons why fog would have lower delays than sunny conditions?

</details>

There's definitely more insights to discover from this real-world dataset. Feel free to continue exploring here, or [download the source dataset](https://figshare.com/articles/dataset/A_high-speed_railway_network_dataset_from_train_operation_records_and_weather_data/15087882/4), which contains several more months of high-speed rail data!